# EMA Project Diary / Lab Notebook

This note book will follow the data expporation and analysis of the schools dataset.

Let's start by looking at the keystage 2 results data.

In [1]:
 !head -5 'data/2015-2016/england_ks2final.csv'

In [2]:
# let's look at how many lines of data there are
!wc -l 'data/2015-2016/england_ks2final.csv'

16316 data/2015-2016/england_ks2final.csv


From the head we can see there are a number of columns and what appears to be some missing data on some lines.  There are over 16000 rows of data.

Let's have a quick look at the KS4 data as well.

In [3]:
 !head -5 'data/2015-2016/england_ks4final.csv'

In [4]:
# let's look at how many lines of data there are
!wc -l 'data/2015-2016/england_ks4final.csv'

5489 data/2015-2016/england_ks4final.csv


The KS4 data looks to have more missing data than the KS2 data, indicated by the number of `,`.  At almost 5500 lines it is a far smaller dataset.

I'll have a quick look at both files in open refine to see if it needs cleaning.

# TODO - OpenRefine

# Import into MongoDB

Due to the size of the datasets I will import them into MongoDB to help with the organisation and access to the data during the investigation.

In [5]:
# import required libraries
import pymongo
import bson

import pandas as pd

In [6]:
# import the KS2 results data
!/usr/bin/mongoimport --port 27351 --drop --db schools --collection ks2results \
    --type csv --headerline --ignoreBlanks \
    --file data/2015-2016/england_ks2final.csv

2018-05-11T16:50:09.280+0000	connected to: localhost:27351
2018-05-11T16:50:09.281+0000	dropping: schools.ks2results
2018-05-11T16:50:12.279+0000	[##############..........] schools.ks2results	10.1MB/16.5MB (61.3%)
2018-05-11T16:50:14.397+0000	[########################] schools.ks2results	16.5MB/16.5MB (100.0%)
2018-05-11T16:50:14.398+0000	imported 16316 documents



That looks good the database has 16316 documents as expected from the line count.  Let's check they look ok.

In [7]:
# import the KS4 results data
!/usr/bin/mongoimport --port 27351 --drop --db schools --collection ks4results \
    --type csv --headerline --ignoreBlanks \
    --file data/2015-2016/england_ks4final.csv

2018-05-11T16:50:14.533+0000	connected to: localhost:27351
2018-05-11T16:50:14.533+0000	dropping: schools.ks4results
2018-05-11T16:50:16.520+0000	imported 5489 documents


That looks good too, 5489 documents as expected.

In [8]:
# open a connection to the mongo server
client = pymongo.MongoClient('mongodb://localhost:27351')

In [9]:
# open the imported database and collections
db = client.schools
ks2 = db.ks2results
ks4 = db.ks4results

In [10]:
# double check that we have all the documents
ks2.find().count(), ks4.find().count()

(16316, 5489)

In [11]:
ks2.find_one()

{'BELIG': 302432,
 'DIFFN_MATPROG': 0,
 'DIFFN_READPROG': 0,
 'DIFFN_RWM_EXP': 0,
 'DIFFN_RWM_HIGH': 0,
 'DIFFN_WRITPROG': 0,
 'GELIG': 289836,
 'GPS_AVERAGE': 104,
 'GPS_AVERAGE_FSM6CLA1A': 102,
 'GPS_AVERAGE_H': 110,
 'GPS_AVERAGE_L': 94,
 'GPS_AVERAGE_M': 103,
 'GPS_AVERAGE_NotFSM6CLA1A': 105,
 'MATPROG_B': 0,
 'MATPROG_B_LOWER': 0,
 'MATPROG_B_UPPER': 0,
 'MATPROG_EAL': 0,
 'MATPROG_EAL_LOWER': 0,
 'MATPROG_EAL_UPPER': 0,
 'MATPROG_FSM6CLA1A': 0,
 'MATPROG_FSM6CLA1A_LOWER': 0,
 'MATPROG_FSM6CLA1A_UPPER': 0,
 'MATPROG_G': 0,
 'MATPROG_G_LOWER': 0,
 'MATPROG_G_UPPER': 0,
 'MATPROG_H': 0,
 'MATPROG_H_LOWER': 0,
 'MATPROG_H_UPPER': 0,
 'MATPROG_L': 0,
 'MATPROG_L_LOWER': 0,
 'MATPROG_L_UPPER': 0,
 'MATPROG_M': 0,
 'MATPROG_MOBN': 0,
 'MATPROG_MOBN_LOWER': 0,
 'MATPROG_MOBN_UPPER': 0,
 'MATPROG_M_LOWER': 0,
 'MATPROG_M_UPPER': 0,
 'MATPROG_NotFSM6CLA1A': 0,
 'MATPROG_NotFSM6CLA1A_LOWER': 0,
 'MATPROG_NotFSM6CLA1A_UPPER': 0,
 'MAT_AVERAGE': 103,
 'MAT_AVERAGE_FSM6CLA1A': 101,
 'MAT_AVERA

In [12]:
ks4.find_one()

{'AC5EM13': '0%',
 'AC5EM14_PTQ': '0%',
 'AC5EM15_PTQ_EE': '0%',
 'AC5EM16_PTQ_EE': '0%',
 'ADDRESS1': 'Queen Victoria Street',
 'AGERANGE': '10-18',
 'ALPHAIND': 11828,
 'ATT8SCR': 42.1,
 'ATT8SCREBAC': 22.2,
 'ATT8SCREBAC_FSM6CLA1A': 'NP',
 'ATT8SCREBAC_NFSM6CLA1A': 'NP',
 'ATT8SCRENG': 7.3,
 'ATT8SCRENG_FSM6CLA1A': 'NP',
 'ATT8SCRENG_NFSM6CLA1A': 'NP',
 'ATT8SCRMAT': 0,
 'ATT8SCRMAT_FSM6CLA1A': 'NP',
 'ATT8SCRMAT_NFSM6CLA1A': 'NP',
 'ATT8SCROPEN': 12.6,
 'ATT8SCROPENG': 10.4,
 'ATT8SCROPENG_FSM6CLA1A': 'NP',
 'ATT8SCROPENG_NFSM6CLA1A': 'NP',
 'ATT8SCROPENNG': 2.2,
 'ATT8SCROPENNG_FSM6CLA1A': 'NP',
 'ATT8SCROPENNG_NFSM6CLA1A': 'NP',
 'ATT8SCROPEN_FSM6CLA1A': 'NP',
 'ATT8SCROPEN_NFSM6CLA1A': 'NP',
 'ATT8SCR_15': 'NA',
 'ATT8SCR_AV': 'NP',
 'ATT8SCR_BOYS': 42.1,
 'ATT8SCR_EAL': 'NP',
 'ATT8SCR_FSM6CLA1A': 'NP',
 'ATT8SCR_GIRLS': 'NA',
 'ATT8SCR_HI': 'NP',
 'ATT8SCR_LO': 'NP',
 'ATT8SCR_NFSM6CLA1A': 'NP',
 'ATT8SCR_NMOB': 'NP',
 'BPUP': 139,
 'CONTFLAG': 0,
 'DIFFN_ATT8': 'NP',
 'DIFFN_

there are a lot of abbreviations it would be useful to have a way to easily access them.  Let's look for an abbreviations file.

In [13]:
!ls data/2015-2016/

abbreviations.xlsx	    england_swf.csv
abs_meta.csv		    england_vaqual.csv
census_meta.csv		    england_vasubj.csv
england_abs.csv		    ks2_meta.csv
england_census.csv	    ks4_meta.csv
england_cfrfull.xlsx	    ks4_meta_methodology.csv
england_ks2final.csv	    ks4-pupdest_meta.csv
england_ks4final.csv	    ks5_meta.csv
england_ks4-pupdest.csv     ks5-studest_meta.csv
england_ks4underlying.xlsx  la_and_region_codes_meta.csv
england_ks5final.csv	    sixth_form_centres_and_consortia_meta.xlsx
england_ks5-studest.csv     spine_meta.csv
england_ks5underlying.xlsx  swf_meta.csv
england_spine.csv


In [23]:
!head data/2015-2016/ks2_meta.csv

Ah, it is in an xlsx file.  I'll need to convert it into a csv file before I can import it to mongodb.

I opened it in excel and then saved it as a csv file.

In [21]:
!head -5 data/2015-2016/abbreviations.csv

In [ ]:
still not great it needs to be transposed I think.  I'll pop it i '